### PR TITLE
Feature/e2e tests workflow

### PR DESCRIPTION
**Pull request by carstingaxion**
This PR allows to run automated and manual end-to-end tests, while sharing the same and existing, `wp-env` based, setup.

## NEEDED, before merge

- [ ] Update all blocks to [API Version](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-api-versions/) 3, to be able to use `editor.canvas` easily. 
(https://github.com/GatherPress/gatherpress/pull/776#issuecomment-2284078360)

## Automated tests 

Check the results of the _e2e-tests action workflow_ at  `https://github.com/GatherPress/gatherpress/actions/workflows/e2e-tests.yml`.

## Manual testing

0. Have `node` installed
1. Have `docker` running
2. Open the plugin folder in a terminal
3. _Choose one of the following options_
   - Run Playwright in the background using
      ```
      npm run test:e2e
      ```
       
   - Run Playwright visually (and change what's happening) using
      ```
      npm run test:e2e:debug
      ```
      grafik

   - Run Tests independently _AND_ visually using the [Playwright VSCode extension](https://playwright.dev/docs/getting-started-vscode)
 

### Description of the Change

Closes #

### Changelog Entry

> Changed existing end-to-tests to rely on wordpress/e2e-test-utils-playwright & setup GitHub actions workflow to run tests using the existing wp-env environment

### Credits

Props @carstingaxion , @swissspidy 

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/GatherPress/gatherpress/blob/main/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

### CODE REVIEW

File: .github/workflows/e2e-tests.yml [hunk 1]
```diff
@@ -1,39 +1,62 @@
...
+                npm run wp-env start
 
-      - name: Playwright install
-        run: npx playwright install --with-deps
+        - name: Running the tests
+          env:
+              WP_AUTH_STORAGE: artifacts/storage-states/admin.json
+              WP_BASE_URL: http://localhost:8889
+              WP_USERNAME: admin
+              WP_PASSWORD: password
```
Comment (swissspidy) [inline .github/workflows/e2e-tests.yml:53]: Unless you have a hyper custom setup you don't need these, these are all the defaults

File: test/e2e/global-setup.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/**
```
Comment (swissspidy) [inline test/e2e/global-setup.ts:1]: Why add your own global-setup? The one provided by Gutenberg does handle all the defaults for you

**Comment by carstingaxion** [comment]
Hello @swissspidy , nice that you take the time & have a look; Thank you!

I've removed the `globalSetup` & the `env` variables, like you said. I'm absolutely fine removing as lot, as possible from not needed overhead! 

While there are many tests failing now, because of my changes; let's ignore most of them for now. Those are the existing tests, that I haven't touched yet. 

But let us please focus on `gatherpress-blocks-editor.spec.ts` which has two tests only and should help to illustrate the new things related to the e2e-test-utils-playwright package.

File: .github/workflows/e2e-tests.yml [hunk 3]
```diff
@@ -1,39 +1,62 @@
...
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          WP_ADMIN_PASSWORD: ${{ secrets.WP_ADMIN_PASSWORD }}
-        run: npm run test:e2e
-        if: ${{ success() || failure() }}
+        - name: Retain failed test results
+          uses: actions/upload-artifact@v3
+          if: failure()
+          with:
+            name: test-results-matrix-less-debugging
+            path: test-results/
```
Comment (swissspidy) [inline .github/workflows/e2e-tests.yml:62]:
```suggestion
            path: artifacts/test-results/
```

**Comment by swissspidy** [comment]
> But let us please focus on `gatherpress-blocks-editor.spec.ts` which has two tests only and should help to illustrate the new things related to the e2e-test-utils-playwright package.
> 
> EDIT: Please take a look at [GatherPress/gatherpress/actions/runs/10348428519/job/28640652754?pr=776#step:8:231](https://github.com/GatherPress/gatherpress/actions/runs/10348428519/job/28640652754?pr=776#step:8:231) which is the result of one of the simplest paragraph tests from gutenberg, which fails.

Well have you tried running them locally with `--debug`? What do you see?

If you fix the artifacts upload on CI then you can also get traces and screenshots from the failed attempt and analyze them locally.

**Comment by carstingaxion** [comment]
> Well have you tried running them locally with --debug? What do you see?

grafik

```bash

carsten@ms-6845:~/Dokumente/webs__at-devilbox/gatherpress/repositories/gatherpress$ npx playwright test --debug --config test/e2e/playwright.config.ts

Running 3 tests using 1 worker

  ✓  1 [chromium] › gatherpress-admin-settings.spec.ts:12:9 › GatherPress Settings › A link to the plugin settings page is present under the Events menu (7.7s)
  ✓  2 [chromium] › gatherpress-blocks-editor.spec.ts:39:2 › GatherPress general block tests › Are 1 blocks available? (9.8s)
Failed to load resource: the server responded with a status of 404 (Not Found)
  ✘  3 [chromium] › gatherpress-blocks-editor.spec.ts:60:2 › GatherPress general block tests › Does the block insert? (2.8m)
Testing stopped early after 1 maximum allowed failures.


  1) [chromium] › gatherpress-blocks-editor.spec.ts:60:2 › GatherPress general block tests › Does the block insert? 

    Error: locator.evaluate: Target page, context or browser has been closed
    Call log:
      - waiting for frameLocator('[name="editor-canvas"]').locator(':root')


      74 |              const firstBlockTagName = await editor.canvas
      75 |                      .locator( ':root' )
    > 76 |                      .evaluate( () => {
         |                       ^
      77 |                              return document.querySelector( '[data-block]' ).tagName;
      78 |                      } );
      79 |

        at /home/carsten/Dokumente/webs__at-devilbox/gatherpress/repositories/gatherpress/test/e2e/tests/gatherpress-blocks-editor.spec.ts:76:5

    attachment #1: trace (application/zip) ─────────────────────────────────────────────────────────
    artifacts/test-results/gatherpress-blocks-editor--e90a3-ests-Does-the-block-insert--chromium/trace.zip
    Usage:

        npx playwright show-trace artifacts/test-results/gatherpress-blocks-editor--e90a3-ests-Does-the-block-insert--chromium/trace.zip

    ────────────────────────────────────────────────────────────────────────────────────────────────

  1 failed
    [chromium] › gatherpress-blocks-editor.spec.ts:60:2 › GatherPress general block tests › Does the block insert? 
  2 passed (3.1m)
  1 error was not a part of any test, see above for details
```

**Comment by swissspidy** [comment]
A screenshot of the browser is more helpful than one from the terminal, but thanks :)

The test you copied from Gutenberg does not work in your use case because your editor is not iframed.

The post editor will be iframed if all registered blocks have a Block API version 3 or higher and the editor has no classic meta boxes below the blocks.

This is the case in the Gutenberg test environment, but not in yours, because all the GatherPress blocks are using apiVersion 2.

So if you are interacting with `editor.canvas, you need to adjust your locator usage accordingly because `editor.canvas` isn't an iframe, so `.locator( ':root' )` won't find anything. Alternatively of course bump your apiVersion to ensure the editor is iframed.

**Comment by carstingaxion** [comment]
> A screenshot of the browser is more helpful than one from the terminal, but thanks :)

I know, sorry. But I didn't knew, what kind of screenshot would make sense.

> The test you copied from Gutenberg does not work in your use case because your editor is not iframed.
> 
> 
> 
> The post editor will be iframed if all registered blocks have a Block API version 3 or higher and the editor has no classic meta boxes below the blocks.
> 
> 
> 
> This is the case in the Gutenberg test environment, but not in yours, because all the GatherPress blocks are using apiVersion 2.
> 
> 
> 
> So if you are interacting with `editor.canvas, you need to adjust your locator usage accordingly because `editor.canvas` isn't an iframe, so `.locator( ':root' )` won't find anything. Alternatively of course bump your apiVersion to ensure the editor is iframed.

This was (I guess) the holy-grail answer, I was looking for, for days now! Thank you, very much @swissspidy !!!

File: .github/workflows/e2e-tests.yml (hunk 1 above)
Comment (carstingaxion) [reply .github/workflows/e2e-tests.yml]: Removed. Thank you.

File: .github/workflows/e2e-tests.yml (hunk 3 above)
Comment (carstingaxion) [reply .github/workflows/e2e-tests.yml]: Fixed. Thanks, too.

File: test/e2e/global-setup.ts (hunk 2 above)
Comment (carstingaxion) [reply test/e2e/global-setup.ts]: Is also removed. Works good, even without it.

**Pull request closed without merging**